boostorg/serialization
# Fix and test singleton implementation

As described in #104 the singleton implementation was broken in https://github.com/boostorg/serialization/commit/b0a794da38b52c8ae0f8155cf3bb30321cf67f00 which led to crashes especially in the context of (multiple) shared libraries.

The "fix" allocating the instance on the heap basically just never freed the instances and hence leaked memory. This was due to not understanding the root cause of the problem.

To summarize:   
There are 2 use cases for the singleton: On where the type `T` is derived from `singleton<T>` and one where `singleton<T>` is used without deriving `T` from it. The above commit put the `is_destroyed` handling into the ctor/dtor of `singleton<T>`. However if `T` is not derived from it, then `singleton<T>` is never instantiated and hence `is_destroyed` is never set. See also some discussion in #104 and #79.

The crash/memory corruption now happens (or happened before the leaky implementation) if the `tkmap` used from `extended_type_info_typeid` is destroyed first. As it used from the ctor of `extended_type_info_typeid` this should normally not happen. However it does happen when used from multiple shared libraries under (at least) GCC 5.4 and 7.1. The reason is currently unknown (see https://stackoverflow.com/questions/50064617/order-of-destruction-for-static-function-members-in-shared-libraries) but we can just assume, the order of destruction is undefined. Hence the need for the `is_destroyed` flag.

On this PR:

- It is built on top of #94 to enable more testing
- It also tests in release mode to ensure that e.g. stripping of symbols by the linker does not cause problems
- Add a test case to ensure, that singletons are properly constructed, destroyed and the `is_destroyed` flag is correctly set
- Add a test case to test the same for shared libraries
- Add a test case based on https://gist.github.com/Flamefire/286e9e0e501731a04f10786450d3e711 to test the issue with multiple shared libraries. Before this PR this caused a double-free which is detected by the debug-glibc. It would be even better to run this in valgrind, but I'm not familiar enough with the Jam-Files and windows doesn't have it.
- Fix the failing testcases by partially reverting https://github.com/boostorg/serialization/commit/b0a794da38b52c8ae0f8155cf3bb30321cf67f00 :
- Reintroduce `detail:singleton_wrapper` with the correct `is_destroyed` setting in the dtor. It uses a method-local variable to enforce initialization on first use (preferred way to avoid static init disaster https://isocpp.org/wiki/faq/ctors#construct-on-first-use-v2) 
- Replace heap-allocated variable by static instance (see isocpp FAQ, avoids static init disaster and is thread safe)
- Assert `!is_destroyed` on enter of `get_instance`. Note that the comment was wrong. It did not refer to the instance and is even removed in non-debug builds
- Make the ctor of `singleton` protected and default to avoid accidental misuse (it might never be called if `T` is not derived from `singleton<T>`)
- Excessively comment everything to avoid breaking changes by misunderstandings

Other notes:

- The comment on the heap allocated variable was wrong: as `singleton_wrapper` is the most derived type of the instance it will always be destructed before `T` and `singleton<T>` (if `T` is derived from it) There is nothing one can do about it!
- Valgrind tests would be nice
- static instance is superior to static pointer (see isocpp FAQ), it is not about visibility, double-destruction or anything else. It simply makes sure that the instance is actually destructed at program exit
- Again: It is safe to assume the destruction order is undefined. Hence all asserts assuming otherwise have been removed (they had a runtime check afterwards anyway which is the correct way)
- Not deriving from `singleton<T>` but using it directly is potentially dangerous as you can have multiple `T` instances and you have only 1 `singleton<T>` where you might not expect it. Example: `extended_type_info` uses a `singleton<detail::tkmap>` to work around this issue because `extended_type_info_typeid` already uses a `singleton<tkmap>`. It might be wise to enforce deriving from `singleton<T>` and using `T::get_instance()` or at least do that in the library.

Finally: Please try to break this! I'm pretty sure I covered everything by using TDD but I might have missed something.
